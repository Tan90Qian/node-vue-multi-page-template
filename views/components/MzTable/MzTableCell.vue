<script>
import { isValidElement } from '../_util/prop-util.js';

function isInvalidRenderCellText(text) {
  return (
    text && !isValidElement(text) && Object.prototype.toString.call(text) === '[object Object]'
  );
}

export default {
  name: 'MzTableCell',
  props: {
    record: Object,
    column: Object,
    index: Number
  },
  methods: {
    handleClick(e) {
      const {
        record,
        column: { onCellClick }
      } = this;
      if (onCellClick) {
        onCellClick(record, e);
      }
    }
  },
  render() {
    const { column, record, index } = this;
    const { dataIndex, customRender, className = '' } = column;
    let text;
    let colSpan;
    let rowSpan;
    const tdProps = {
      attrs: {},
      class: className || column.class,
      on: {
        click: this.handleClick
      }
    };
    if (typeof dataIndex === 'string' || typeof dataIndex === 'number') {
      text = record[dataIndex];
    } else if (!dataIndex || dataIndex.length === 0) {
      text = record;
    }
    if (customRender) {
      text = customRender.call(this, text, record, index);
      if (isInvalidRenderCellText(text)) {
        tdProps.attrs = text.attrs;
        colSpan = tdProps.attrs.colSpan;
        rowSpan = tdProps.attrs.rowSpan;
        text = text.children;
      }
    }
    if (isInvalidRenderCellText(text)) {
      text = null;
    }
    if (rowSpan === 0 || colSpan === 0) {
      return null;
    }
    if (column.align) {
      tdProps.style = { textAlign: column.align };
    }
    if (column.width) {
      tdProps.attrs.width = column.width;
    }
    return <td {...tdProps}>{text}</td>;
  }
};
</script>
