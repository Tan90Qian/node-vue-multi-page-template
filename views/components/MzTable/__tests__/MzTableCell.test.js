/* eslint-disable no-unused-vars */
import { mount } from '@vue/test-utils';
import sinon from 'sinon';

import MzTableCell from '../MzTableCell';

describe('MzTableCell tests', () => {
  let propsData;
  let record;
  beforeEach(() => {
    record = {
      a: 1
    };
    propsData = {
      column: {
        dataIndex: 'a'
      },
      record,
      index: 0
    };
  });
  it('should render simple td by default', () => {
    const wrapper = mount(MzTableCell, {
      propsData
    });
    const td = wrapper.find('td');

    expect(td.exists()).toBeTruthy();
    expect(td.text()).toBe('1');
  });

  it('should add class name from column', () => {
    propsData.column.className = 'test';
    const wrapper = mount(MzTableCell, {
      propsData
    });
    const td = wrapper.find('td');

    expect(td.classes('test')).toBeTruthy();
  });

  it('should set text-align style when align in column', () => {
    propsData.column.align = 'right';
    const wrapper = mount(MzTableCell, {
      propsData
    });
    const td = wrapper.find('td');

    expect(td.html()).toContain('text-align: right');
  });

  it('should set width attribute when width in column', () => {
    propsData.column.width = 300;
    const wrapper = mount(MzTableCell, {
      propsData
    });
    const td = wrapper.find('td');

    expect(td.attributes('width')).toBe('300');
  });

  it('should render content by customRender function', () => {
    propsData.column.customRender = function customRender(val, record, index) {
      const h = this.$createElement;
      return <span class="test">{`${val} - ${index}`}</span>;
    };
    const wrapper = mount(MzTableCell, {
      propsData
    });
    const td = wrapper.find('td');

    expect(td.html()).toContain('<span class="test">1 - 0</span>');
  });

  it('should render null when rowSpan or colSpan equal to 0', () => {
    propsData.column.customRender = function customRender(val, record, index) {
      const h = this.$createElement;
      return {
        children: <span>test</span>,
        attrs: {
          rowSpan: 0
        }
      };
    };
    const wrapper = mount(MzTableCell, {
      propsData
    });
    const td = wrapper.find('td');

    expect(td.exists()).toBeFalsy();
  });

  it('should call onCellClick when td trigger click event', () => {
    propsData.column.onCellClick = (record, e) => {
      console.log(record, e);
    };
    const onCellClickMock = sinon
      .mock(propsData.column)
      .expects('onCellClick')
      .withArgs(record);
    const wrapper = mount(MzTableCell, {
      propsData
    });
    const td = wrapper.find('td');
    td.trigger('click');

    onCellClickMock.verify();
  });
});
