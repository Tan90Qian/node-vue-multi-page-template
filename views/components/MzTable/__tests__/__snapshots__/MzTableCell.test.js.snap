// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MzTableCell tests renders correctly 1`] = `
<td>
  1
</td>
`;
